### ISSUE TITLE
[BUG] AttributeError in SubLOF with novelty=False when calling fit_predict

### ISSUE THREAD

**Issue by axisrow** [sktime/sktime#7652]
When using SubLOF from the sktime library with novelty=False and calling fit_transform or fit_predict, an AttributeError is raised. The error indicates that the predict method is not available when novelty=False, which contradicts the documentation's recommendation to use fit_predict for outlier detection on the training data.

**To Reproduce**
```python
import pandas as pd
from sktime.annotation.lof import SubLOF
model = SubLOF(3, window_size=5, novelty=False)
x = pd.DataFrame([0, 0.5, 100, 0.1, 0, 0, 0, 100, 0, 0, 0.3, -1, 0, 100, 0.2])
model.fit_transform(x)
```

**Expected behavior**

I expected the fit_predict method to detect outliers in the provided dataset x without raising an AttributeError. According to the documentation, fit_predict should be used for outlier detection when novelty=False.

From the documentation:

> By default, LocalOutlierFactor is only meant to be used for outlier
>         detection (novelty=False). Set novelty to True if you want to use
>         LocalOutlierFactor for novelty detection. In this case be aware that
>         you should only use predict, decision_function and score_samples
>         on new unseen data and not on the training set; and note that the
>         results obtained this way may differ from the standard LOF results.

**Versions**
<details>

System:
    python: 3.10.11 (v3.10.11:7d4cc5aa85, Apr  4 2023, 19:05:19) [Clang 13.0.0 (clang-1300.0.29.30)]
executable: /usr/local/bin/python3
   machine: macOS-15.2-arm64-arm-64bit

Python dependencies:
          pip: 24.3.1
       sktime: 0.35.0
      sklearn: 1.5.2
       skbase: 0.12.0
        numpy: 1.26.4
        scipy: 1.14.1
       pandas: 2.2.3
   matplotlib: 3.9.3
       joblib: 1.4.2
        numba: 0.60.0
  statsmodels: 0.14.4
     pmdarima: None
statsforecast: 2.0.0
      tsfresh: None
      tslearn: None
        torch: 2.5.1
   tensorflow: 2.16.2
</details>

**Comment by KrishBakshi** [comment]
Hi there! I will be working towards resolving this issue.

**Comment by fkiraly** [comment]
@KrishBakshi, thanks!

I think we need to do two things:

* ensure the `_fit` etc have lower case `y`, in `SubLOF`, this can be a single PR
* regarding the issue with `novelty`, I do not see how `predict` would work in this case, since the `sklearn` estimator does not provide this. @Alex-JG3 , as the author of the estimator, what do you think?

**Comment by KrishBakshi** [comment]
Hi @fkiraly  I have fixed the `Y` typo and created a draft PR, Do you have any idea how to fix the `novelty = False` issue?

**Comment by Alex-JG3** [comment]
### Context

`SubLOF` uses sklearn's `LocalOutlierFactor` algorithm under the hood which has two modes: outlier detection (`novelty=False`), and novelty detection (`novelty=True`). Information about the difference between the two modes can be found [here](https://scikit-learn.org/stable/modules/outlier_detection.html). To summaries, outlier detection mode should only be used when when the training data is polluted with anomalies and the goal is the to find anomalies in the training data. Because this should not be used to find anomalies in new observations, `predict` is not available.

### Possible solution

In `SubLOF`, `fit_transform(x)` is almost equivalent to `fit(x).predict(x)` under the hood which is why we get the attribute error. `fit_predict` is available for `LocalOutlierFactor` when `novelty=False` so we could override `fit_transform` and change it to use `fit_predict(x)` instead of `fit(x).predict(x)`. I think that would solve this issue.

**Comment by Harshavardhan-28** [comment]
Agree with the mode summary, the sklearn docs are clear that outlier mode only exposes `fit_predict`.

One thing on the proposed fix though: overriding `fit_transform` would only cover `fit_transform` itself. `fit_predict`, `predict`, `transform`, `predict_points` and `predict_segments` all still land in `_predict`, where `model.predict(...)` is called without checking the mode, so they would keep raising. The draft in #8051 overrides `fit_predict` instead, which does also cover `fit_transform` since `fit_transform` calls it, but the other four are still left out. (Side note, #8051 links #7562 in the body rather than this issue, which is why it doesn't show up in this thread.)

How about handling it in `_predict`, since that's the one place they all funnel through? Roughly: when `novelty=False`, compute the labels in `_fit` using `fit_predict` and store them, then have `_predict` return those. That's close to how sklearn does it anyway, since in outlier mode the scores live on `negative_outlier_factor_` as a fitted attribute rather than being worked out later.

@fkiraly that still leaves the question you raised earlier, what `predict` should do with genuinely new data when `novelty=False`. sklearn has no equivalent there, so raising a clear error seems more honest than quietly refitting, but happy either way.

@alex-gregory-ds if you arent working on #8051 , I'd be glad to work on this and open a PR  with my approach.
